verifiable-db/src/query/batching/row_chunk/consecutive_rows.rs

```diff
@@ -0,0 +1,1290 @@
+        // which should always hold for consecutive rows since:
+        // - if the successor of first row is in range, then second row must be its successor
+        //   in the same rows tree, and so the predecessor of second row is the first row itself,
+        //   which is expected to be in range since we never need to prove nodes not in range
+        //   but with a successor in range
+        // - if the successor of first row is out of range, then second row is expected to
+        //   be a node in the "subsequent" rows tree (i.e., the rows tree stored in the index
+        //   tree node which is the successor of the index tree node storing first row); this node
+        //   can be either:
+        //	    - the first node of the "subsequent" rows tree with value >= min_secondary;
+        //        in this case, the predecessor of second row is < min_secondary, and so out of range
+        //	    - if no such node can be found in the "subsequent" rows tree, then second row will be
+        //        the last node in the "subsequent" rows tree with value < MIN_secondary; in
+        //	      this case, also its predecessor will necessarily be < MIN_secondary, and so
+        //	      out of range
+        // we first compute first_row_succ_in_range XOR second_row_pred_in_range: a XOR b = a + b - 2*a*b
+        let range_flags_sum = b.add(
+            first_node_succ_in_range.target,
+            second_node_pred_in_range.target,
+        );
+        let range_flags_xor = b.arithmetic(
+            F::NEG_ONE + F::NEG_ONE,
+            F::ONE,
+            first_node_succ_in_range.target,
+            second_node_pred_in_range.target,
+            range_flags_sum,
+        );
+        // then, `are_consecutive = are_consecutive AND NOT(range_flags_xor) = are_consecutive - are_consecutive*range_flags_xor`
+        are_consecutive = BoolTarget::new_unsafe(b.arithmetic(
+            F::NEG_ONE,
+            F::ONE,
+            are_consecutive.target,
+            range_flags_xor,
+            are_consecutive.target,
+        ));
+        // in case of nodes in a rows tree, then we need to enforce that second is the successor of first only
+        // if the nodes are in the same rows tree, that is if `first_node_succ_in_range` is true
+        let new_are_consecutive = b.and(are_consecutive, either_is_in_path);
+        are_consecutive = BoolTarget::new_unsafe(b.select(
+            first_node_succ_in_range,
+            new_are_consecutive.target,
+            are_consecutive.target,
+        ));
+    }
+
+    (are_consecutive, first_node_succ_in_range)
+}
+
+pub(crate) fn are_consecutive_rows(
+    b: &mut CBuilder,
+    first: &BoundaryRowDataTarget,
+    second: &BoundaryRowDataTarget,
+    min_primary: &UInt256Target,
+    max_primary: &UInt256Target,
+    min_secondary: &UInt256Target,
+    max_secondary: &UInt256Target,
```

Comment:
can you precise whether those are from the query or just from a node ? I assume the query but would be nice to have clear context all the time when talking about min and max.

Reply:
Renamed in commit https://github.com/Lagrange-Labs/mapreduce-plonky2/pull/396/commits/5787dc75a5d6df1a467910e6fc2b524e21943460